File: euler/python/P102.py
"""
Triangle Containment
Problem 102
"""

TRIANGLES_TXT = "P102.txt"
POINTS = 3
ORIGIN = (0, 0)


triangles = []

# Read the file as a set of tuples (i.e. each point)
with open(TRIANGLES_TXT, 'r') as f:
    for i in map(lambda x: map(int, x[:-1].split(",")), f.readlines()):
        triangles.append([(i[2 * j], i[2 * j + 1]) for j in xrange(POINTS)])

# For read-ability, the first index of 2-element tuples are refered to as x,
# and the second as y
x, y = 0, 1

# For each triangle, make sure the ORIGIN is on the same side of each line
# as the third point in question. We represent side as a boolean.
def side_of(a, b, c):
    if b[x] == c[x]:
        return a[x] > b[x]
    slope = (b[y] - c[y]) / float(b[x] - c[x])
    return a[y] > a[x] * slope + c[y] - slope * c[x]

# Returns true if the ORIGIN is contained by the triangle
def contained(a, b, c):
    return side_of(a, b, c) == side_of(ORIGIN, b, c) and \
           side_of(b, a, c) == side_of(ORIGIN, a, c) and \
           side_of(c, a, b) == side_of(ORIGIN, a, b)

print sum(contained(*i) for i in triangles)

File: euler/ruby/P102.rb
"""
  Triangle Containment
  Problem 102
"""

TRIANGLES_TXT = "P102.txt"
DIMENSIONS = 2
ORIGIN = [0] * DIMENSIONS
DELTA = 0.000001

# We first format the file into arrays of triangles.
triangles = File.read(TRIANGLES_TXT).split("\n")
triangles.map!{ |n| n.split(",").map(&:to_i).each_slice(DIMENSIONS).to_a }

# Returns the area of the triangle defined by the three given points. We do this via Heron's formula
def triangle_area(*sides)
  lengths = (-1...DIMENSIONS)
            .map{ |i| Math.sqrt((0...DIMENSIONS)
            .map{ |j| (sides[i][j] - sides[i - 1][j]) ** 2 }.inject(:+)) }
  Math.sqrt(lengths.push(0).map{ |i| lengths.inject(:+) / Float(2) - i }.inject(:*))
end

# Checks the area of each triangle formed by two points and the origin, and the triangle itself.
# If the areas are equal, then the origin is bounded by the triangle. Because floats suck, we only
# get an approximate equality using DELTA.
puts triangles.select{ |i| (i.combination(DIMENSIONS).to_a.map{ |j| triangle_area(*j, ORIGIN) }
                             .inject(:+) - triangle_area(*i)).abs < DELTA }.count
